# Coursework 2, 2026

## Instructions

**Please read** the assessment brief on the ELE page for details on completing the assignment, submission, marking guidelines, and AI guidance.

You should work through the background material in the file `CW2 Helper File.ipynb`.

You **must not** import any functions other than those in the cell below. These are functions from the **finite fields**,  **crypto basic**, and **math** modules.

**Ensure** you execute the following cell to import the required functions.

In [1]:
from finite_fields import *
from crypto_basic import crt
from math import sqrt, ceil, prod

## Question 1

This question focuses on hand-calculation of inverses in finite fields via polynomial long division and the extended Euclidean algorithm for polynomials. Read the section *Manual Finite Field Calculations* in the `CW2 Helper Field` before starting this question.

## Question 1(a) [3 marks]

This question should be answered **by hand** and typset below, except that you **may** use Python or an electronic calculator to compute values of the polynomial $m$.

Let $K$ be a field. Recall that for a polynomial $p \in K[X]$ to be *irreducible* means that $p$ is not a constant and if $p = ab$, where $a, b \in K[X]$, then either $a$ or $b$ are constant polynomials.

Let $m \in \mathbb F_7[X]$ be the polynomial given by $m(X) = 3X^3 + 2X + 1$. Prove that $m$ is irreducible.

Proof:

Let $K = \mathbb(F)_7$,

m is a third order polynomial, the only possible non-trivial factorisation of a third order polynomial is a quadratic and linear factor.

Assuming m is reducible, then if $m = ab$, where $a, b \in K[X]$, then let $a$ be the quadratic polynomial that factors $m$ and $b$ be the linear polynomial that factors $m$.  

So $m = a \times b = (Ax^2 + Bx + C) \times (Dx + E)$. 

From this we can say that $Dx + E$ is a factor of $m$ and thus $-\frac{E}{D}$ is a root of $m$, where $-\frac{E}{D} \in K$ (division is valid here as $b$ is degree one so $D \neq 0$). 

$\therefore \,$  $ m$ is reducible $\implies$ $m$ has a root in $K$

with the contrapositive:

$m$ doesn't have a root in $K$ $\implies$ $m$ is irreducible

So now we can check the elements of $K = \{0, 1, 2, 3, 4, 5, 6\}$, if none are roots of $m$, then $m$ is irreducible


In [2]:
def poly_m(X):
    return (3*X**3 + 2*X + 1) % 7

In [3]:
# test elements of K from 0-6
for X in range(7):
    print(poly_m(X))

1
6
1
4
5
1
3


As no elements of $K$ are roots of $m$ ,  $\therefore \, m$ is irreducible. 

## Question 1(b) [15 marks]

This question is to be completed **by hand**. That is, without using Python or any other computer package for computation, though you may use an electronic calculator. Your answer should be typeset below in $\LaTeX$ using the model presented in `CW2 Helper File.ipynb`.

Let $F = \mathbb F_7[X]/m \mathbb F_7[X]$ be the finite field with $7^3$ elements, where $m$ is the polynomial defined in Question 1(a). Let $f \in \mathbb F_7[X]$ be the polynomial $f(X) = 5 X^2 + X + 3$.

Find a polynomial $g \in \mathbb F_7[X]$ such that $\bar g$ is the multiplicative inverse of $\bar f$ in $F$. Here, $\bar h$ denotes the left coset $\bar h =h + m \mathbb F_7[X]$.

Please add as many markdown cells below as necessary.

### Solving

We start by doing polynomial long division. Let $r_0(X) = m(X) = 3X^3 + 2X + 1$, $r_1(X) = f(X) = 5X^2 + X + 3$. We now find $r_2$ and $q_1$ such that $r_0 = r_1q_1 + r_2$ and $r_2 = 0$ or $deg(r_2) < deg(r_1)$. So $r_0 \div r_1$. 

Note: since we are working in the field $\mathbb{F}_7$, all calculations are done modulo $7$. So for instance the first calculation would be $3 \times 5^{-1} = 3 \times 3 = 9 \equiv 2$. Similarly for all subsequent calculations.  

$$
 \begin{array}{l|rrrr}
                 &      &       & 2X   & 1    \\
            \hline
   5X^2 + X + 3  & 3X^3 &       & 2X   & 1    \\
                 & 3X^3 & 2X^2  & 6X   &      \\
            \hline
                 &      & 5X^2  & 3X   & 1    \\
                 &      & 5X^2  & X    & 3    \\
            \hline
                &       &       & 2X  & 5 
 \end{array}
$$

We find $r_2(X) = 2X + 5$ and $q_1(X) = 2X +1 $.

And again for $r_3$ and $q_2$. So $r_1 \div r_2$.

$$
 \begin{array}{l|rrr}
                                  &            & 6X  & 3   \\
                             \hline
             2X + 5               & 5X^2       &  X  & 3   \\
                                  & 5X^2       & 2X  &     \\
                             \hline
                                  &            & 6X  & 3    \\
                                  &            & 6X  & 1    \\
                             \hline
                                  &            &     & 2
\end{array}
$$

We find $r_3(X) = 2$ and $q_2(X) = 6X + 3$.

Since $r_3$ is a non-zero constant, $\text{gcd}(r_0, r_1) = 1$. So $\bar{f}$ is invertible.

We have the equations: 

$r_0 - r_1q_1 = r_2$, and 

$r_1 - r_2q_2 = r_3$ 

combining them gives:

$r_3 = r_1 - (r_0 - r_1q_1)q_2 = r_1(1 + q_1q_2) - r_0q_2$

as $r_3 = 2$, multiply each side by $2^{-1} = 4$ to get the expression in the form $1 = ...$, which let's us utilise Bezout's identity.

Hence, $ 1 = 4r_1(1 + q_1q_2) - 4r_0q_2$, which can be reduced modulo $r_0 = m$,

to get $ 1 \equiv 4r_1(1 + q_1q_2) \equiv \bar{f} 4(1 + (2X + 1)(6X + 3)) \equiv \bar{f}(48X^2 + 48X + 16) \equiv \bar{f}(6X^2 + 6X + 2) (\text{mod}7)$

So take $g = 6X^2 + 6X +2$ or equivalent modulo $m$. 


---

## Question 2

In this question, you'll be writing a test for irreducibility of polynomials over the field $\mathbb F_2$. Read the sections *Finite Fields in Python* and  *An Irreducibility Test* from the `CW2 Helper File` before starting this question.

## Question 2(a)(i) [5 marks]

Complete the definition of the Python function `divides_frob` below. It takes a non-constant polynomial $f \in \mathbb F_2[X]$ and returns `True` if $f$ divides the polynomial $X^{2^n} - X$, where $n = \deg(f)$. Otherwise it returns `False`.

There is a naïve but highly inefficient way to write the function. Ensure your implementation is efficient.

### Plan:

$f \vert (X^{2^n} - X) \implies X^{2^n} - X \equiv 0 (\text{mod} f) \implies X^{2^n} \equiv X \, (\text{mod} f )$ as it's in $\mathbb{F}_2$, $-X \equiv X (\text{mod} f)$.

So, rather than computing $X^{2^n} - X$ and checking divisibility (bad as $2^n$ is generally large), I can use repeated doubling on $X^{2^n}$ and compare to $X (\text{mod} f)$ as shown above.

In [4]:
def divides_frob(f):
    """Determines if the non-constant polynomial f in GF(2)[X] divides X^{2^n} - X, where n = deg(f).
    Equally, if X^{2^n} = X (mod f)"""
    n = f.degree()
    F = GF2QuotientField(f)
    g = F(0b10)
    for i in range(n):
        g = g**2  # as it's in F it automatically reduces modulo f
    return g == F(0b10)

## Question 2(a)(ii) [2 marks]

Run your function `divides_frob` to determine if the polynomial $f(X) = X^{100}$ divides the polynomial $X^{2^{100}}-X$. If your code fails to complete, you may need to rewrite your answer to Question 2(a)(i).

In [5]:
# Test
f = GF2Polynomial(0b10)**100
divides_frob(f)

False

Which makes sense from the irreducibility test theorem in section 3 of the helper file. It says $f$ is irreducible iff, $f$ divides $X^{q^n}$, where $q = 2$ in this case and $n$ is the degree of $f$ as above. So as $f$ is clearly reducible (eg. $X^{100} = X^{99} \times X$), it will not divide the polynomial.  

## Question 2(b)(i) [4 marks]

Complete the definition of the function `distinct_prime_factors` below so that for each integer $n \ge 2$, `distinct_prime_factors(n)` is a list of the distict prime factors of $n$, listed in order of increasing size. Thus, `distinct_prime_factors(12)` should be `[2, 3]`.

In [6]:
def distinct_prime_factors(n):
    """Gives a list of distinct primes whose product is n"""
    d = 2
    listo = []
    while d*d <= n:
        if n % d == 0:    
            listo.append(d)   # by appending beforehand, I ensure distinct prime factors
        while n % d == 0:
            n = n //d
        d = d + 1
    if n == 1:
        return listo
    else:
        listo.append(n)
        return listo

## Question 2(b)(ii) [1 mark]

Run your function `distinct_prime_factors` to determine the distinct prime factors of 226512 and of 125921.

In [7]:
# Test
print(distinct_prime_factors(226512))
print(distinct_prime_factors(125921))

[2, 3, 11, 13]
[125921]


## Question 2(c)(i) [4 marks]

Suppose $f \in \mathbb F_2[X]$ is a polynomial of degree $n \ge 2$. Suppose $p_1< \dots< p_k$ is a list of the distinct prime factors of $n$. Let $n_i = n / p_i$ for $i=1, \dots, k$.

Complete the definition of the function `all_gcd_frob` below. If $\gcd(f, X^{2^{n_i}}-X) = 1$ for $i=1, \dots, k$, the function returns `True`. Otherwise, it returns `False`.

### Plan:

1. Get the degree of polynomial $f$ and find its distinct prime factors
2. Compute $n_i = \frac{n}{p_i}$ for each prime factor $p_i$
3. Compute $X^{2^n_i} (\text{mod} f)$ as the gcd calculation yields the same result, but now we are working with smaller polynomials
4. Add $X$ (equivalent to subtracting in $\mathbb{F}_2$)
5. Compute $\text{gcd}(f, X^{2^n_i} - X)$ using the gcd function
6. Reutrn `False` if any gcd isn't 1, otherwise return `True` 

In [8]:
def all_gcd_frob(f):
    """Given a polynomial f in GF2[X] of degree n >= 2. Let p_1, ... p_k be the distinct prime factors of n.
    Let n_i = n / p_i (for i = 1, ..., k). This function returns True if gcd(f, X^(2^n_i) - X) = 1 for 
    i = 1, ..., k. Otherwise, it returns False."""
    # 1.
    n = f.degree()
    ps = distinct_prime_factors(n)
    for p in ps:
        # 2.
        ni = n // p
        # 3.
        F = GF2QuotientField(f) 
        g = F(0b10) 
        for i in range(ni):
            g = g**2
        # 4.
        g = g + F(0b10) 
        # 5. & 6. 
        if f.gcd(g.poly) != GF2Polynomial(0b1):
            return False
    return True

## Question 2(c)(ii) [2 marks]

Display the result of applying your function `all_gcd_frob` to the polynomials $X^{100}$ and $X^{100} + 1$. If you find that your function does not complete on either of these polynomials, you may need to write `all_gcd_frob` more efficiently.

In [9]:
# Test
f1 = GF2Polynomial(0b10) ** 100
f2 = f + GF2Polynomial(0b1)
print(all_gcd_frob(f1))
print(all_gcd_frob(f2))

False
True


Once again, from the irreducibility theorem, this makes sense. $X^{100}$ is clearly reducible and it's less obvious for $X^{100} + 1$, but it looks pretty irreducible.

## Question 2(d)(i) [3 marks]

Use your answers to Questions 2(a)(i) and 2(c)(i) to complete the definition of the function `is_irreducible` below. Given a polynomial $f \in \mathbb F_2[X]$ of degree at least $2$, running `is_irreducible(f)` should return `True` if $f$ is irreducible and `False` otherwise.

### Plan:

I know $f$ is irreducible iff
 
* $f$ divides $X^{q^n} - X$ and
* $\gcd(f, X^{q^{n_i}}-X) = 1$, for each $i=1, \dots, k$.

So I need the function to use 2(a)(i) `divides_frob(f)` for the first condition and use 2(c)(i) `all_gcd_frob(f)` for the second condition. If either returns `False`, then the function `is_reducible(f)` returns `False`, otherwise it returns `True`.

In [10]:
def is_irreducible(f):
    """For f in GF2[X] of degree at least 2, returns True when f is irreducible and False otherwise."""
    return divides_frob(f) and all_gcd_frob(f)

## Question 2(d)(ii) [2 marks]

Use your answer to Question 2(d)(i) to write and execute code that lists all the irreducible polynomials in $\mathbb F_2[X]$ of degree 5.

### Thoughts:

Since the polynomials in $\mathbb{F}_2$ are represented as binary strings and they are all degree 5, each string will be of the form `0b1XXXXX`, where each `X` can be a 1 or 0. So I'm looking at binary numbers `0b100000`-`0b111111` which is 32-63. 

In [11]:
irreduc_poly_list = []
for n in range(32,64):
    f = GF2Polynomial(n)
    if is_irreducible(f):
        irreduc_poly_list.append(f)
     
for f in irreduc_poly_list: display(f)

GF2Polynomial(0b100101)

GF2Polynomial(0b101001)

GF2Polynomial(0b101111)

GF2Polynomial(0b110111)

GF2Polynomial(0b111011)

GF2Polynomial(0b111101)

---

## Question 3

This question is about using the baby-step giant-step algorithm to solve the discrete logarithm problem. Ensure you have read the sections *Finite Fields in Python* and *Dictionaries* from the `CW2 Helper File` before beginning this question.

## Question 3(a)(i) [4 marks]

Suppose $s \in \mathbb F_2[X]$ is an irreducible polynomial. Then the ring $\mathbb F_2[X]/s \mathbb F_2[X]$ is a finite field. Justifying your answer, give an expression for the order of this field in terms of the degree of $s$. Do this **by hand** (that is, without using Python).

### Answer:

Let $\text{deg}(s) = n$.

Every element of the ring is a coset of $\bar{r}$, where $r$ is the remainder from dividing by $s$, so $\text{deg}(r) < n$.

Thus, every element can be written uniquely as: $a_0 + a_1 X + a_2 X^2 + ... + a_{n-1} X^{n-1}, \, \, a_i \in \mathbb{F}_2$.

There are $n$ coefficients, each is in $\mathbb{F}_2$, so they are either a $0$ or a $1$. $2$ choices made $n$ times gives $2^n$ elements

## Question 3(a)(ii) [3 marks]

Let `s` be an irreducible polynomial over $\mathbb F_2[X]$. Using Python, the object `GF2QuotientField(s)` represents the quotient field $\mathbb F_2[X]/s\mathbb F_2[X]$.

Suppose `F = GF2QuotientField(s)` is such an object. Then `F.modulus` is the polynomial `s` used to create `F`.

Below, complete the definition of the function `field_order` so that `field_order(F)` returns the order of the finite field `F`, where `F` is an `GF2QuotientField` object

In [12]:
def field_order(F):
    """Returns the order of the finite field F, where F is a GF2QuotientField object"""
    n = F.modulus.degree()
    return 2**n

## Question 3(a)(iii) [0 marks]

Run the cells below to verify that your function `field_order` works as expected.

In [13]:
s1, s2 = GF2Polynomial(0b10011), GF2Polynomial(0b1101)
display(s1, s2)
is_irreducible(s1), is_irreducible(s2)

GF2Polynomial(0b10011)

GF2Polynomial(0b1101)

(True, True)

In [14]:
F1, F2 = GF2QuotientField(s1), GF2QuotientField(s2)
F1.modulus, F2.modulus

(GF2Polynomial(0b10011), GF2Polynomial(0b1101))

In [15]:
field_order(F1), field_order(F2)

(16, 8)

## Question 3(b)(i) [4 marks]

Complete the definition of the following function `is_generator`. It expects as input `f`, an element of a finite field `GF2QuotientField` object. Then `isgenerator(f)` should be True when `f` is a generator of the unit group of the finite field and should be `False` otherwise.

You should use the ideas we saw when discussing the primitive root test.

### Plan:

The primitive root test says: Let $p$ be an odd prime, and $a \in \mathbb{Z}, p \nmid a$. Then $a$ is a primitive root
modulo $p$ if and only if, for each prime factor $q$
of $p − 1, a^{\frac{p−1}{q}} \not\equiv 1 \text{mod} p$. 

If $a$ is a primitive root modulo $p$, then it is a generator of the unit group $(\mathbb{Z}/p\mathbb{Z})^\times$.

So we just need to conduct the primitive root test to check whether f is a generator. 

In [16]:
def is_generator(f):
    """Determines whether f a generator of the unit group of the finite field to which it belongs"""
    if not isinstance(f, GF2FieldElement): # ensure f is an element of a finite field
        raise TypeError("is_generator expects a field element")
    F = f.field # the finite field to which f belongs
    one = F(1) # the one polynomial in the finite field.
    n = field_order(F) - 1
    factor_list = distinct_prime_factors(n)
    for q in factor_list:
        if f**(n // q) == one:
            return False
    return True

## Question 3(b)(ii) [1 mark]

Supppose `F1` is as in Question 3(a)(iii). Let $f$ be the residue class of the polynomial $X^3 + X + 1$ in the field `F1`. Using your function `is_generator` above, determine if `f` is a generator of the unit group of `F1`.

In [17]:
f = F1.element(GF2Polynomial(0b1011))
is_generator(f)

True

## Question 3(b)(iii) [3 marks]

Let `F1` be as in Question 3(a)(iii). Use your function `is_generator` to find all the elements `g` in the finite field `F1` such that `g` is  a generator of the unit group of `F1` and is the residue class of a polynomial of degree 3.

### Thoughts: 

I am considering all elements `g`, in `F1`, and then considering:

1. is it a generator of the unit group
2. is it degree 3

So I can just consider all degree 3 polynomials in $\mathbb{F}_2$, and check if they're generators. There are 8 polynomials to consider as the leading coefficient is $1$ in order for it to be degree 3, and each of the follwoing coefficients has two choices so $2^3$ possibilities. 

In [18]:
for i in range(8, 15): # 0b1000, 0b1111
    g = F1.element(GF2Polynomial(i))
    if is_generator(g):
        display(g)

GF2FieldElement_19(0b1001)

GF2FieldElement_19(0b1011)

GF2FieldElement_19(0b1101)

GF2FieldElement_19(0b1110)

## Question 3(c) [2 marks]

Let $m$ be the polynomial in $\mathbb F_2[X]$ where $m(X) = a_{32}X^{32} + \ldots + a_1 X + a_0$ and where the string $a_{32} a_{32} \dots a_1 a_0$ is the binary representation of the decimal number 4567890193.

Define and display `m` in Python and test if `m` is irreducible.

In [19]:
m = GF2Polynomial(bin(4567890193))
m

GF2Polynomial(0b100010000010001000111100100010001)

In [20]:
is_irreducible(m)

True

## Question 3(d) [6 marks]

Complete the definition of the function `bsgs` below. It should take non-zero arguments `g` and `h` that belong to a finite field (technically, that are instances of `GF2FieldElement`), and belong to the same field. You may assume `g` is a generator of the unit group of the quotient field to which both $g$ and $h$ belong. The function returns a non-negative integer `z` such that $g^z = h$.

Your function should implement the baby-step giant-step method. I suggest creating a dictionary for the baby step list where the key is the group element and the value is the exponent. You can save time and space by not storing the giant list. Instead, iterate through the elements of the list, checking if each element is in the baby list.

### Plan:

1. Set $n = \lceil \sqrt{N} \rceil$, smallest integer that is at least $\sqrt{N}$, where $N = \text{ord}(g) = 2^{\text{degree}} -1$, as $g$ is a generator of the unit group 
2. Construct a list $e, g, g^2, ..., g^{n-1}$.
3. Construct a list $h, hg^{-n}, hg^{-2n},..., hg^{-(n-1)n}$. But as negative powers won't work, I can use Fermat's little theorem: since $g^N = 1$ in the group $\implies g^{-1} = g^{N-1} \implies g^{-n} = g^{N - n}$, which is non-negative as $n \leq N$.
4. Search for a match between the lists. That is, find $0 \leq i < n$ and $0 \leq j < n$ such that $g^i = hg^{-jn}$.
5. Then $h = g^{i+jn}$ so $x \leftarrow i + jn$ is a solution.

In [21]:
def bsgs(g, h):
    """Given that g and h are both non-zero elements of the same finite field, this function uses
    baby-step giant-step to return z such that g^z = h, if such a z exists"""
    if not isinstance(g,  GF2FieldElement) or not isinstance (h, GF2FieldElement):
        raise TypeError(f"{g} and {h} must be field elements")
    F = g.field
    if F != h.field:
        raise ValueError(f"{g} and {h} must belong to the same field")
    if g == 0 or h == 0:
        raise ValueError(f"both arguments {g} and {h} must be non-zero")

    # 1
    deg = g.field.modulus.degree()
    N = 2**deg - 1
    n = ceil(sqrt(N))
    
    # 2
    baby_dict = {}
    for i in range(n):
        baby_dict[g**i] = i  # baby step

    # 3
    current_giant = h
    giant_step = g**(N-n) # g^(-n) = g^(N-n)
    for j in range(n):    
        # 4 check for match   
        if current_giant in baby_dict:
            return baby_dict[current_giant] + j*n # 5
        current_giant = current_giant * giant_step  # giant step
     
    return None

### Test:

In [22]:
f = irreduc_poly_list[3]
F = GF2QuotientField(f)
g = F(0b10)  # generator g = X, don't know this yet, just found with trial by error

In [23]:
F.element(3)

GF2FieldElement_55(0b11)

In [24]:
F(3)

GF2FieldElement_55(0b11)

In [25]:
print(bsgs(g, g**0))
print(bsgs(g, g**1))
print(bsgs(g, g**15))
print(bsgs(g, g**30))
print(bsgs(g, g**60)) # = 26 (mod31) "since it's a 5th degree polynomial so 2^5 elements and 31 in the unit group

0
1
15
30
29


## Question 3(e)(i) [3 marks]

Complete the function `get_generator` below. It takes a field $F$ and returns a generator of the unit group $F^\times$. Your function should call the function `is_generator` defined in Question 3(b)(i).

### Plan:

I already have a function to determine whether an element of a field is a generator, so I just need to loop over elements of F and return the first generator I find.

In [26]:
def get_generator(F):
    """Returns a generator of the unit group of the field F"""
    if not isinstance(F, GF2QuotientField):
        raise TypeError(f"{F} must be a GF2QuotientField")

    size = field_order(F)
    # iterate over elements of F 
    for i in range(1, size):
        f = F(i)
        if is_generator(f):
            return f

    raise ValueError("No generator found — is F a valid field?")

## Question 3(e)(ii) [2 marks]

In Question 3(c), you should have showed that the polynomial $m$ is irreducible. Hence, the quotient ring $F = \mathbb F_2[X]/m\mathbb F_2[X]$ is a field.

Use your function `get_generator` from Question 3(e)(i) to write and execute code to find and display a generator of $F^\times$. Assign this generator to the variable `g`.

In [27]:
F = GF2QuotientField(m)
g = get_generator(F)
g

GF2FieldElement_4567890193(0b11)

## Question 3(f) [2 marks]

Using the $g$ you produced in your answer to question 3(e)(ii), write and execute code to find and display an integer $z$ such that $g^z = \overline{X^{11}+ X^5 + 1}$. Verify your answer.

In [28]:
h = F(bin(2**11+2**5+1))
h

GF2FieldElement_4567890193(0b100000100001)

In [29]:
z = bsgs(g,h)
print(z)
display(g**z)

286994260


GF2FieldElement_4567890193(0b100000100001)

## Question 4

This question is about the Pohlig-Hellman algorithm for solving the discrete logarithm problem. Before starting this question, ensure you have read the sections *Finite Fields in Python*, *Dictionaries*, *Passing Functions as Arguments to Functions*, and *Using the `crt` function* from the `CW2 Helper File`. You shoud complete Question 3 before starting this question.


## Question 4(a) [3 marks]

In Question 3, you wrote a baby-step giant-step algorithm to solve the DLP $g^x =h$ where you assumed that $g$ is a generator of the unit group of the finite field to which both $g$ and $h$ belong. The algorithm is not so efficient if $g$ and $h$ belong to some proper subgroup of this group.

Below, complete the definition of the function `bsgs_sub` where `bsgs_sub(g, h, N)` solves the same discrete logarithm problem but where `g` is assumed to be a generator of an order `N` subgroup of the unit group of the finite field and where `h` is assumed to be an element of this subgroup. This should involve only a small modification of your previous code.

Add an appropriate docstring (the string between triple quotation marks) to describe the function.

In [30]:
def bsgs_sub(g, h, N):
    """Given that g and h are both non-zero elements of the same finite field and N is assumed to be 
    the order of the subgroup generated by g, this function uses baby-step giant-step to return z such that 
    g^z = h, if such a z exists. Returns None if no such z exists."""
    if not isinstance(g,  GF2FieldElement) or not isinstance (h, GF2FieldElement):
        raise TypeError(f"{g} and {h} must be field elements")
    F = g.field
    if F != h.field:
        raise ValueError(f"{g} and {h} must belong to the same field")
    if g == 0 or h == 0:
        raise ValueError(f"both arguments {g} and {h} must be non-zero")
    if not isinstance(N, int) or N < 1:
        raise ValueError("N must be a positive integer")

    # 1
    n = ceil(sqrt(N))
    
    # 2
    baby_dict = {}
    for i in range(n):
        baby_dict[g**i] = i  # baby step

    # 3
    current_giant = h
    giant_step = g**(N-n) # g^(-n) = g^(N-n)
    for j in range(n):    
        # 4 check for match   
        if current_giant in baby_dict:
            return baby_dict[current_giant] + j*n # 5
        current_giant = current_giant * giant_step  # giant step
     
    return None

## Question 4(b)(i) [4 marks]

Let $n \ge 2$ be an integer. Complete the definition of the function `prime_power_factors` below such that `prime_power_factors(n)` is a list of all the prime power factors of $n$, in increasing order.

For example, `prime_power_factors(200)` should be `[8, 25]`. You may factorise via trial division.

In [31]:
def prime_power_factors(n):
    """Gives a list of the prime power factors of n"""
    d = 2
    dicto = {}
    while d*d <= n:
        while n % d == 0:
            if d not in dicto:
                dicto[d] = 0
            dicto[d] += 1
            n = n //d
        d = d + 1
    if n == 1:
        return [p**k for p, k in dicto.items()]
    else:
        dicto[n] = 1
        return [p**k for p, k in dicto.items()]

## Question 4(b)(ii) [2 marks]

Use your function `prime_power_factors` to produce and display the prime power factorisations of 49, 163, and 5040.

The `prod` function takes a list and returns the product of the elements in the list. Use `prod` 
to check your answers.

In [32]:
print(prime_power_factors(49))
print(prime_power_factors(163))
print(prime_power_factors(5040))

print(prod(prime_power_factors(49)))
print(prod(prime_power_factors(163)))
print(prod(prime_power_factors(5040)))

[49]
[163]
[16, 9, 5, 7]
49
163
5040


## Question 4(c)(i) [6 marks]

Complete the definition of the the function `pohlig_hellman` below. Here, `f` is a discrete log solving algorithm such that `f(a, b, N)` returns `x` such that $a^x = b$ where $a$ and $b$ belong to some group of order $N$ and where $a$ is a generator of the group.

You should use the `crt` function from the module `crypto_basic`, which is available in the file `crypto_basic.py`.

### Plan:

1. Find group order
2. Find prime power factors
3. Set $g_i = g^{N/q_{i}^{e_i}}$, for $1 \leq i \leq r$
4. Set $h_i = h^{N/q_{i}^{e_i}}$, for $1 \leq i \leq r$
5. For each $1, ..., r$, set $G_i = \langle g_i \rangle$
6. Solve DLP $g_{i}^{x_i} = h_i$.
7. Finally, using the Chinese Remainder Theorem, solve: $x \equiv x_1 (\text{mod} \, q_{1}^{e_1}), \,  ... \,, \,  x \equiv x_1 (\text{mod} \, q_{r}^{e_r})$ 

In [33]:
def pohlig_hellman(g, h, f):
    """Given that g and h are both non-zero elements of the same finite field, this function uses
    Pohlig-Hellman to return z such that g^z = h, if such a z exists. The argument f is a function such that
    f(a, b, N) returns x for which a^x = b, where a is a generator of a group of order N that contains b"""

    # 1 - Find group order
    deg = g.field.modulus.degree()
    N = 2**deg - 1

    #2 - Find prime power factors
    qs = prime_power_factors(N)  # each of these is already given to it's respective power

    # 3 and 4 - Set gi and hi
    gs, hs = [], [] 
    g_prime, h_prime = g, h
    for q in qs:
        g = g_prime**(N // q)
        gs.append(g)
        h = h_prime**(N // q)
        hs.append(h)

    # 5 - set Gi

    # 6 - solve DLP for gi ** xi = hi, f is the solving algorithm
    xs = []
    for gi, hi, q in zip(gs, hs, qs):
        xs.append(f(gi, hi, q))
    
    # 7 - CRT
    return crt(xs, qs)    

## Question 4(c)(ii) [2 marks]

Run your function `pohlig_hellman` with `g` as in youur answer to Question 3(e)(ii). Take `h` to be $g^{100}$. Take `f` to be the function `bsgs_sub` that you wrote in your answer to Question 4(a).

In [34]:
F = GF2QuotientField(m)
g = get_generator(F)
h = g**100
print(pohlig_hellman(g, h, bsgs_sub))

100


## Question 4(d) [2 marks]

We'll investigate the discrete logarithm problem in a much larger group than that in Question 3.

Let $t$ be the polynomial in $\mathbb F_2[X]$ where $t(X) = a_{64}X^{64} + \ldots + a_1 X + a_0$ and where the string $a_{64} a_{64} \dots a_1 a_0$ is the binary representation of the decimal number 26799019411345421523.

Write and execute Python code to test if $t$ is irreducible.

In [35]:
t = GF2Polynomial(bin(26799019411345421523))
is_irreducible(t)

True

## Question 4(e) [2 marks]

As $t$, from Question 4(d), is irreducible, the quotient ring $F = \mathbb F_2[X]/ t\mathbb F_2[X]$ is a field. Use your function `get_generator` from Question 3(e)(i) to find a generator of the unit group $F^\times$ and assign it to the variable `g`. Display `g`.

In [36]:
F = GF2QuotientField(t)
g = get_generator(F)
g

GF2FieldElement_26799019411345421523(0b110)

## Question 4(f) [2 marks]

With $F$ as in Question 4(e), let $h \in F$ be a field element that is represented by the polynomial  
$b_{31} X^{31} + b_{30} X^{30} + \dots + b_1 X + b_0 \in \mathbb F_2[X]$ where $b_{31} \dots b_1 b_0$ is the binary representation of the number which has decimal representation `2840568370`.

Write and run Python code to find and display a non-negative integer $z$ such that $g^z = h$, where $g$ is the generator of $F$ that you found in your answer to Question 4(e). Verify your answer.

In [37]:
h = F(bin(2840568370))
z = pohlig_hellman(g, h, bsgs_sub)
display(z)

11656435251555532932

## Question 4(g) [6 marks]

Discuss whether the Pohlig-Helmman method of the baby-step giant-step method is more efficient for the problem in Question 4(f). Give order-of-magnitude estimates for the number of steps (counted as group multiplications) required in justifying your answer. You should write your answer **by hand** but you may use Python to gather and analyse data needed for your answer. You should not use any code timing or profiling tools (like `timeit` or `cProfile`).

### Thinking:

For each method, considering the number of group multiplications required as the efficiency gives: 

* The BSGS method has complexity:

A baby step loop of size $n = \lceil \sqrt{N} \rceil$ and giant step loop also of size $n = \lceil \sqrt{N} \rceil$

So that gives $\mathcal{O}(2 \sqrt{N})$ = $\mathcal{O}(\sqrt{N})$

* The Pohlig-Helmman method has complexity:

BSGS is run once for each prime power factor $q_i$ of $N$, in which it calls on the subgroup of order $q_i$, so costs $\mathcal{O}(\sum_{i=1}^r\sqrt{q_i})$  

In Question 4(f), I am considering the quotient field $\mathbb{F}_2/ \langle t \rangle$, where $t$ is a degree 64 irreducible polynomial.

For the BSGS, I have that $N$ is calculated as follows:     
``` Python
    deg = g.field.modulus.degree()
    N = 2**deg - 1
```
This gives:

In [38]:
N = 2**64 - 1
print(N)
print(sqrt(N))

18446744073709551615
4294967296.0


hence $N = 2^{64} - 1 \approx 1.8 \times 10^{19} \implies \sqrt{N} \approx 4 \times 10^{9}$, so it's around  $\mathcal{O}(10^9)$ complexity.

For the Pohlig-Helmman, I have to consider the prime power factors of N, as that gives the $q_i$ values

In [39]:
qs = prime_power_factors(N)
qs

[3, 5, 17, 257, 641, 65537, 6700417]

So now I can calculate the $\sum_{i=1}^r\sqrt{q_i}$:

In [40]:
sum([sqrt(q) for q in qs])

2893.958745305533

So the complexity is around: $ \mathcal{O}(10^3)$

### Conclusion:

So for this problem where $N = 2^64 - 1$, BSGS requires $\mathcal{o} (\sqrt{N}) \approx 4 \times 10^9$ group multiplications, while the Pohlig-Helmman method requires $\mathcal{O} (\sum_{i=1}^{r} \sqrt{q_i}  \approx 2894$ group multiplications.

$\therefore$ the Pohlig-Helmman method is significantly more efficient, by a factor of approximately $10^6$.